### ISSUE TITLE
Define `_MDSPAN_NO_UNIQUE_ADDRESS` based on CXX version.

### ISSUE THREAD

**Issue by trivialfis** [kokkos/mdspan#161]
Hi, we run into an issue with nvcc 1.6 and g++ 9.4 using c++ 17. The macro `_MDSPAN_NO_UNIQUE_ADDRESS` is defined differently by these two compilers. I opened a fix https://github.com/rapidsai/raft/pull/764, do you think it's a viable approach and would you like to have it in here as well?

**Comment by mhoemmen** [comment]
@trivialfis Thanks for reporting this and for proposing a fix!  I'll comment on the PR link you posted; thanks!

**Comment by crtrott** [comment]
yeah seems ok to me

**Comment by crtrott** [comment]
Need to check whether testing passes with that though. Also its not that the macro would be differently defined, I think its just that one of them does apply it correct and the other doesn't.

**Comment by crtrott** [comment]
It may also be worthwhile to check whether it works correct with something like NVCC 11.7 and GCC 11 as the host compiler, since the no_unique_address business does help reduce register pressure I would assume.

**Comment by crtrott** [comment]
Also I am not 100% sure that it would fully work, whats the minimum GCC version for C++20 mode with mdspan? If it is like GCC 9.2 then it might still not work with nvcc+gcc in C++20 mode

**Comment by trivialfis** [comment]
> whats the minimum GCC version for C++20 mode with mdspan

This test failed with g++ 9.4 https://github.com/kokkos/mdspan/blob/d56cb92d27a78be0e64efc7788697af1f91ae695/compilation_tests/ctest_no_unique_address.cpp#L96 but succeeded with g++ 11.2. I think g++ 11 is required to pass it. For both g++ 9.4 and g++ 10.3, the `__cpluscplus` macro is defined as `201709` when `-std=c++20/-std=c++2a` is specified, while g++11.2 defines the value as `202002`.

**Comment by trivialfis** [comment]
https://github.com/kokkos/mdspan/blob/d56cb92d27a78be0e64efc7788697af1f91ae695/include/experimental/__p0009_bits/config.hpp#L138 This might also cause troubles. Since nvcc is usually used in combination with other compilers, I think making specialization for it is quite dangerous.

**Comment by mhoemmen** [comment]
@trivialfis wrote:
> Since nvcc is usually used in combination with other compilers, I think making specialization for it is quite dangerous.

I appreciate your caution, but @crtrott has been using nvcc in combination with other compilers for quite a while now : - ) .  We'll certainly take a look at this.

**Comment by achirkin** [comment]
> I appreciate your caution, but crtrott has been using nvcc in combination with other compilers for quite a while now : - ) . We'll certainly take a look at this.

I don't mean to doubt you expertise, but this statement looks a bit funny in the context of the issue that has bugged several projects in a very obscure way :-)

**Comment by mhoemmen** [comment]
@trivialfis FYI, the changes here broke builds for other compilers.  @crtrott 's PR https://github.com/kokkos/mdspan/pull/171 fixes that.  Could you please check whether it works for you?  Thanks! : - )

**Comment by crtrott** [comment]
> > whats the minimum GCC version for C++20 mode with mdspan
> 
> This test failed with g++ 9.4
> 
> https://github.com/kokkos/mdspan/blob/d56cb92d27a78be0e64efc7788697af1f91ae695/compilation_tests/ctest_no_unique_address.cpp#L96
> 
> but succeeded with g++ 11.2. I think g++ 11 is required to pass it. For both g++ 9.4 and g++ 10.3, the `__cpluscplus` macro is defined as `201709` when `-std=c++20/-std=c++2a` is specified, while g++11.2 defines the value as `202002`.

This particular issue was with NVCC and g++9.4 right?

**Comment by trivialfis** [comment]
Hi @crtrott , the test failed with g++9/10, nvcc is not involved here.

**Comment by crtrott** [comment]
What specific config fails? And do you mean with #171 or before that? I did test #171 with GCC 9.1/10.3/11.1 in C++ 14/17/20 mode and GCC 11.1 in C++23 mode.

**Comment by trivialfis** [comment]
Before this issue (161). Your PR looks good, thank you for following up on this.

**Comment by crtrott** [comment]
I also just rerun the version prior to merging #165 and don't see any failures

**Comment by crtrott** [comment]
I see, as I said I can't reproduce that either. Can you send me the exact config you did to run into a problem with pure GCC? Just in case I am overlooking something so I can put that into my test harness.

**Comment by crtrott** [comment]
Mine was something like this: 
```
cmake -DMDSPAN_ENABLE_TESTS=ON -DMDSPAN_ENABLE_BENCHMARKS=ON -DCMAKE_CXX_FLAGS='-Wall -Wextra -Werror' -DCMAKE_CXX_STANDARD=14 -DMDSPAN_CXX_STANDARD=14 -DCMAKE_CXX_COMPILER=g++ -DCMAKE_CXX_EXTENSIONS=OFF /ascldap/users/crtrott/Kokkos/mdspan
```

and for 20 adding -pedantic to the flags.

**Comment by trivialfis** [comment]
Apologies for the ambiguity.  That test needed a fix in #165 because in #165 I changed the configuration of `no_unique_address`. So, it only breaks if we merge #165 without the fix in that test. It should always work in the stable branch.

I left the comment about that test in order to provide some context on why I need to change it to avoid confusion for reviewers.

**Comment by crtrott** [comment]
Ah ok I get it now. Yeah it was still broken in C++17 mode even with GCC-11. But With #171 everything should be fine now including your original issues I gather.
